### PR TITLE
Don't force allocation of the entire `Stream::buffered` buffer up front

### PR DESCRIPTION
**Pull request by khuey**
This is problematic for large buffer sizes, in particular `Stream::buffered(usize::max_value())` does not do what one might expect it to do.  `VecDeque` is largely a drop in replacement here.

Given how much marketing of futures-rs emphasizes the "zero allocation" bit (which, without language features like `impl Trait`, I think is oversold) I'm not sure how you'll feel about this.  We can add a separate method if necessary, I suppose.

CC @rocallahan

### CODE REVIEW

**Comment by khuey** [comment]
There's also another big problem with having large amounts of buffering: every `Buffered::poll` polls every future it holds, not just the one that's ready.  The solution to that seems to be in flux (`with_unpark_event` is deprecated) so I don't propose to solve that here.

**Comment by alexcrichton** [comment]
Thanks for the PR! If you're working with large numbers of futures in `Buffered` though then yeah I'd definitely recommend working through the with_unpark_event-like solutions. I think that using `stream::FuturesUnordered` (the pseudo-replacement of `with_unpark_event`) may actually be pretty easy here, would you be interested in doing that? That should solve the allocation problem in addition to the "this always polls everything" problem.

**Comment by khuey** [comment]
I do actually want the results in order though.

**Comment by alexcrichton** [comment]
Right yes of course, my thinking is that:

* When a future is pushed into the queue, you push it into the internal `FuturesUnordered` queue
* Each future pushed in also gets a ticket assigned to it
* When futures pop out you check to see if the ticket is the next up
  * If so, you return the future
  * If not, you buffer the result in something like a priority queue

That way you'd get all the perf benefits and should retain minimal memory footprint I think.

**Comment by khuey** [comment]
Hmm.  I'll look into that tomorrow.

**Comment by rocallahan** [comment]
Maybe `FuturesUnordered` should not describe itself as a queue, since its elements are not ordered from the user's point of view. It's really a set.

**Comment by Stebalien** [comment]
In general, it's best to avoid working with a potentially unbounded (or very large) number of futures. This is a recipe for a dos-by-oom vulnerability.

Additionally, futures are zero-allocation in a steady-state so, personally, I'd *expect* it to do all of its allocation up-front. This avoids repeated re-allocations under heavy load (precisely when you don't want such allocations). 

Finally, if you correctly tune the buffer size, you'll end up allocating the full buffer at some point anyways (and tokio will never call `shrink_to_fit` on the buffer to release memory) so you might as well do this up front as it makes your program's memory profile more predictable.

**Comment by rocallahan** [comment]
It's useful to be able to put an upper bound on memory usage without requiring all that memory to be allocated in cases where it won't be needed.

For complex applications with highly variable workloads it's not always possible to "correctly tune" a fixed buffer size.

**Comment by rocallahan** [comment]
In other words, for some applications it makes sense to have "predictable memory usage" by making it always use the worst-case amount of memory. But for other applications it doesn't.

**Comment by alexcrichton** [comment]
> Maybe FuturesUnordered should not describe itself as a queue, since its elements are not ordered from the user's point of view. It's really a set.

[Excellent point](https://github.com/alexcrichton/futures-rs/issues/505)!

> Finally, if you correctly tune the buffer size, you'll end up allocating the full buffer at some point anyways

@Stebalien We also ran into https://github.com/tokio-rs/tokio-core/issues/124 which was fixed by https://github.com/tokio-rs/tokio-core/pull/126, so it's not 100% always the best strategy to preallocate up front from what I can tell (you may not know whether you actually reach the limit or not.

> Hmm. I'll look into that tomorrow.

@khuey ok! Let me know if there's any difficulties though, I can also help look into it and whip up an implementation.

**Comment by khuey** [comment]
Something like that, perhaps, with `Stream::buffered` implemented on top of it?

We may end up with `buffered_unordered`/`buffered_ordered`/`buffered_fixed` or similar to support all three options.

**Comment by alexcrichton** [comment]
Looks good to me, thanks! And yeah I'd imagine that we could easily build a `Stream::buffered` on top of this `FuturesOrdered` implementation. What do you mean by `buffered_ordered` and `buffered_fixed`? We've actually already got [`buffer_unordered`](https://github.com/alexcrichton/futures-rs/blob/master/src/stream/buffer_unordered.rs) which uses `FuturesUnordered` internally

**Comment by khuey** [comment]
By `buffer_fixed` I mean the requirements @Stebalien is talking about.

**Comment by khuey** [comment]
Although, since `FuturesUnordered` will allocate internally, that really means what we have today and not merely `FuturesOrdered` but with a `BinaryHeap::with_capacity` up front.

**Comment by alexcrichton** [comment]
Ah right yes! We can probably hold off on that for now and wait for it to be motivated, I'm ok starting out just having `Stream::buffered` and `Stream::buffer_unordered`

**Comment by khuey** [comment]
This is ready to be reviewed.

File: src/stream/mod.rs [hunk 1]
```diff
@@ -880,7 +882,7 @@ pub trait Stream {
     /// This method is only available when the `use_std` feature of this
     /// library is activated, and it is activated by default.
     #[cfg(feature = "use_std")]
-    fn buffered(self, amt: usize) -> Buffered<Self>
+    fn buffered(self, amt: usize) -> Buffered<FuturesOrdered<<<Self as Stream>::Item as IntoFuture>::Future>, Self>
```
Comment (alexcrichton) [inline src/stream/mod.rs]: Thanks for the updates! Sorry for the delay just got back from vacation.

I think this is a breaking change though, right? Would it be possible to land this without breaking changes? I think it's fine to have a bit of duplication here to preserve these existing type signatures perhaps?

Comment (khuey) [inline src/stream/mod.rs]: I think the return types specifically could be papered over with a type alias.  The `FuturesSet` trait introduction is harder.  But does `FuturesUnordered` actually see a much direct use where consumers `::new()` one and then `push` to it manually?

Comment (alexcrichton) [inline src/stream/mod.rs]: In any case I'd prefer to keep all the separate types for future flexibility. The internals can look like this (sharing impls) but the API should continue to expose unique types.

**Comment by alexcrichton** [comment]
Ok I've touched up a few extra pieces here and there and merged, thanks again!

**Pull request merged**